internal/gps/result.go

```diff
@@ -62,103 +64,63 @@ func WriteDepTree(basedir string, l Lock, sm SourceManager, sv bool, logger *log
 		return fmt.Errorf("must provide non-nil Lock to WriteDepTree")
 	}
 
-	err := os.MkdirAll(basedir, 0777)
-	if err != nil {
+	if err := os.MkdirAll(basedir, 0777); err != nil {
 		return err
 	}
 
+	g, ctx := errgroup.WithContext(context.TODO())
 	lps := l.Projects()
+	sem := make(chan struct{}, concurrentWriters)
+	var cnt uint32
 
-	type resp struct {
-		i   int
-		err error
-	}
-	respCh := make(chan resp, len(lps))
-	writeCh := make(chan int, len(lps))
-	cancel := make(chan struct{})
-
-	// Queue work.
 	for i := range lps {
-		writeCh <- i
-	}
-	close(writeCh)
-	// Launch writers.
-	writers := concurrentWriters
-	if len(lps) < writers {
-		writers = len(lps)
-	}
-	var wg sync.WaitGroup
-	wg.Add(writers)
-	for i := 0; i < writers; i++ {
-		go func() {
-			defer wg.Done()
-
-			for i := range writeCh {
-				select {
-				case <-cancel:
-					return
-				default:
+		p := lps[i] // per-iteration copy
+
+		g.Go(func() error {
+			sem <- struct{}{}
+			defer func() { <-sem }()
+
+			err := func() error {
+				if err := ctx.Err(); err != nil {
+					return err
 				}
 
-				p := lps[i]
-				to := filepath.FromSlash(filepath.Join(basedir, string(p.Ident().ProjectRoot)))
+				ident := p.Ident()
+				projectRoot := string(ident.ProjectRoot)
+				to := filepath.FromSlash(filepath.Join(basedir, projectRoot))
 
-				if err := sm.ExportProject(p.Ident(), p.Version(), to); err != nil {
-					respCh <- resp{i, errors.Wrapf(err, "failed to export %s", p.Ident().ProjectRoot)}
-					continue
+				if err := sm.ExportProject(ident, p.Version(), to); err != nil {
+					return errors.Wrapf(err, "failed to export %s", projectRoot)
 				}
 
 				if sv {
-					select {
-					case <-cancel:
-						return
-					default:
+					if err := ctx.Err(); err != nil {
+						return err
 					}
 
 					if err := filepath.Walk(to, stripVendor); err != nil {
-						respCh <- resp{i, errors.Wrapf(err, "failed to strip vendor from %s", p.Ident().ProjectRoot)}
-						continue
+						return errors.Wrapf(err, "failed to strip vendor from %s", projectRoot)
 					}
 				}
 
-				respCh <- resp{i, nil}
-			}
-		}()
-	}
-	// Monitor writers
-	go func() {
-		wg.Wait()
-		close(respCh)
-	}()
-
-	// Log results and collect errors
-	var errs []error
-	var cnt int
-	for resp := range respCh {
-		cnt++
-		msg := "Wrote"
-		if resp.err != nil {
-			if len(errs) == 0 {
-				close(cancel)
+				return nil
+			}()
+
+			msg := "Wrote"
+			if err != nil {
+				msg = "Failed to write"
 			}
-			errs = append(errs, resp.err)
-			msg = "Failed to write"
-		}
-		p := lps[resp.i]
-		logger.Printf("(%d/%d) %s %s@%s\n", cnt, len(lps), msg, p.Ident(), p.Version())
-	}
+			logger.Printf("(%d/%d) %s %s@%s\n", atomic.AddUint32(&cnt, 1), len(lps), msg, p.Ident(), p.Version())
```

Comment:
Are these technically racing to log _after_ the atomic add (meaning they could reorder)?  Would the op have to be crammed inside a `fmt.Formatter` to get around that?

Reply:
Yeah, they might reorder. I'm not sure fmt.Formatter would do it, because there's still a race in the machinery that ends up calling Write with the result.

We could do it under our own lock, though.

Reply:
What about moving the `sem` stuff into the body of `err func()`? Then another lock here is not such a big deal.

Reply:
I don't follow. What does the semaphore have to do with anything?

Reply:
Then another worker is freed _before_ we potentially lock for this count, or even now before the logger locks internally.